Tasks/task3

alexandr-irza opened the pull request:
First version of task 3

alexandr-irza committed:
Added scripts

Added creation scripts
Added filling scripts

alexandr-irza committed:
Fixed files encoding

alexandr-irza committed:
Completed query for task

alex-geraschenko commented on task3/data/1_Categories.sql:
```diff
@@ -0,0 +1,11 @@
+﻿DELETE Categories WHERE ParentId IS NOT NULL; --Remove sub categories
```
Дополнительный `DELETE` не нужен

alex-geraschenko commented on task3/data/1_Categories.sql:
```diff
@@ -0,0 +1,11 @@
+﻿DELETE Categories WHERE ParentId IS NOT NULL; --Remove sub categories
+DELETE Categories; --Remove root categories
```
Если этот скрипт запустить когда вся база заполнена, то будет ошибка. Лучше добавить еще один скрипт в самом начале, который будет в нужном порядке очищать таблицы, а потом отдельными скриптами заполнять таблицы

alexandr-irza replied:
При заполнении данных было проще иметь удаление в каждом файле. Вынес удаление в отдельный файл.

alex-geraschenko commented on task3/data/4_RecipeIngredients.sql:
```diff
@@ -0,0 +1,20 @@
+﻿DELETE RecipeIngredients; --Remove exist data
```
Когда нет внешних ключей, можно использовать `TRUNCATE TABLE`

alex-geraschenko commented on task3/data/5_RecipeSteps.sql:
```diff
@@ -0,0 +1,10 @@
+﻿DELETE RecipeSteps; --Remove exist recipes
```
Когда нет внешних ключей, можно использовать `TRUNCATE TABLE`

alex-geraschenko commented on task3/queries/FirstThreeRecipes.sql:
```diff
@@ -0,0 +1,9 @@
+from ( select top 3  r.Id, r.Name, c.Id as CategotyId, c.Name as CategoryName
+	from Recipes r
+	inner join Categories c on r.CategoryId = c.Id
+	where c.Id = 3 or c.ParentId = 3
```
Этот запрос не сработает, если будет больше 1-го уровня вложености категорий. Посмотрите в сторону **Common Table Expressions**

alexandr-irza replied:
Читал об этом, но не учел что может быть больше 2 уровней.

alexandr-irza committed:
Fixed feedbacks

Moved delete opearations to separate file
Rewrote script for selecting first 3 recipes to use more than 2 category levels

alexandr-irza committed:
Complete fixing feedbacks

alex-geraschenko commented on task3/queries/FirstThreeRecipes.sql:
```diff
@@ -0,0 +1,24 @@
+from ( select top 3 r.Id, r.Name, c.CategoryId, c.CategoryPath
+	from Recipes r
+	inner join Categories_CTE c on c.CategoryId = r.CategoryId
+	where c.CategoryPath like '3\%'
```
Лучше не стоит использовать поиск по подстроке на вычисляемых полях. Это может сказаться на производительности при больших объемах данных. Постарайтесь найти решение с использованием индексированых полей (в частности `Id`)

alexandr-irza committed:
Changed script to do not use 'LIKE'

alex-geraschenko commented on task3/queries/FirstThreeRecipes.sql:
```diff
@@ -0,0 +1,27 @@
+	select c.Id, c.ParentId, c.Name, CategoryPath + cast(c.Name as nvarchar(max)) + '\'
+	from Categories c
+	inner join Categories_CTE cc on cc.CategoryId = c.ParentId
+	where cc.CategoryId = @CategoryId or cc.CategoryParentId = @CategoryId
```
Условие `where` избыточно

alexandr-irza replied:
Перенес условие в джоин

alex-geraschenko replied:
Вы не так поняли, его стоит совсем убрать

alexandr-irza committed:
Removed 'WHERE'

alex-geraschenko commented on task3/queries/FirstThreeRecipes.sql:
```diff
@@ -0,0 +1,26 @@
+	union all
+	select c.Id, c.ParentId, c.Name, CategoryPath + cast(c.Name as nvarchar(max)) + '\'
+	from Categories c
+	inner join Categories_CTE cc on cc.CategoryId = c.ParentId and (cc.CategoryId = @CategoryId OR cc.CategoryParentId = @CategoryId)
```
Условие из `WHERE` нужно было не переносить, а убрать совсем. Оно не имеет смысла, т.к. у Вас рекурсивный `CTE`

alexandr-irza replied:
Уже понял, мы используем результаты предыдущего выполнения, а он уже имеет нужный фильтр

alexandr-irza committed:
Removed 'WHERE' and changed scripts for creating tables

alex-geraschenko reviewed (approved):
Хорошо, можно мержить
